### PR TITLE
Support incomplete shape inference for slice_channel and convolution

### PR DESCRIPTION
**Pull request by sxjscience**
Useful for convolutional rnn + GRU

### CODE REVIEW

File: src/operator/convolution-inl.h [hunk 1]
```diff
@@ -445,10 +463,6 @@ class ConvolutionProp : public OperatorProperty {
         << "incorrect stride size: " << param_.stride;
       CHECK_GT(param_.dilate.Size(), 0) \
         << "incorrect dilate size: " << param_.dilate;
-      CHECK(ksize_d < dshape[2] + 2 * param_.pad[0]
```
Comment (piiswrong) [inline src/operator/convolution-inl.h]: why remove this?

Comment (piiswrong) [inline src/operator/convolution-inl.h]: ok i see

File: tests/python/unittest/test_infer_shape.py [hunk 2]
```diff
@@ -44,7 +44,47 @@ def test_backward_infer():
...
+    assert out_shapes[0] == (5,)
+
+
+def test_incomplete_infer_convolution():
+    a = mx.sym.zeros(shape=(0, 10, 0, 0))
+    b = mx.sym.Convolution(data=a, num_filter=21, kernel=(3, 3), dilate=(1, 1), pad=(1, 1))
+    c = mx.sym.zeros(shape=(5, 21, 32, 32))
+    d = b + c
+    arg_shapes, out_shapes, aux_shapes = d.infer_shape()
+    assert out_shapes[0] == (5, 21, 32, 32)
```
Comment (piiswrong) [inline tests/python/unittest/test_infer_shape.py]: shouldn't you check input shape here?

File: src/operator/convolution-inl.h [hunk 3]
```diff
@@ -461,6 +475,33 @@ class ConvolutionProp : public OperatorProperty {
       oshape[4] = (dshape[4] + 2 * param_.pad[2] -
           (1 * (ksize_x - 1) + 1)) / param_.stride[2] + 1;
       SHAPE_ASSIGN_CHECK(*out_shape, 0, ConvertLayout(oshape, kNCDHW, param_.layout.value()));
+      if ((*out_shape)[0].ndim() != 0) {
```
Comment (piiswrong) [inline src/operator/convolution-inl.h]: It's always != 0 here since you just assigned to it

File: src/operator/convolution-inl.h [hunk 4]
```diff
@@ -417,6 +414,27 @@ class ConvolutionProp : public OperatorProperty {
       oshape[3] = (dshape[3] + 2 * param_.pad[1] -
           (param_.dilate[1] * (ksize_x - 1) + 1)) / param_.stride[1] + 1;
       SHAPE_ASSIGN_CHECK(*out_shape, 0, ConvertLayout(oshape, kNCHW, param_.layout.value()));
+      if ((*out_shape)[0].ndim() != 0) {
```
Comment (piiswrong) [inline src/operator/convolution-inl.h]: always !=0

File: tests/python/unittest/test_infer_shape.py (hunk 2 above)
Comment (sxjscience) [inline tests/python/unittest/test_infer_shape.py]: I find that arg_shapes will not include the zeros.
```python
a = mx.sym.zeros(shape=(10, 10))
b = mx.sym.zeros(shape=(10, 10))
c = a+b
c.infer_shape()
```
Returns
```text
([], [(10, 10)], [])
```

Comment (piiswrong) [inline tests/python/unittest/test_infer_shape.py]: Use variable instead of zeros

**Comment by sxjscience** [comment]
@piiswrong I've revised the code accordingly.

File: src/operator/slice_channel-inl.h [hunk 5]
```diff
@@ -172,15 +173,35 @@ class SliceChannelProp : public OperatorProperty {
       << ") does not divide input dimension "
       << real_axis << " (" << dshape[real_axis] << ").";
     dshape[real_axis] /= param_.num_outputs;
-    if (param_.squeeze_axis && dshape[real_axis] == 1) {
+    if (param_.squeeze_axis && (dshape[real_axis] == 1 || ishape[real_axis] == 0)) {
```
Comment (piiswrong) [inline src/operator/slice_channel-inl.h]: Cannot squeeze when ishape[real_axis] == 0.
when squeeze_axis=true and ishape[real_axis] == 0 then ndim of output cannot be decided.

Comment (sxjscience) [inline src/operator/slice_channel-inl.h]: I think that when squeeze_axis=true, ishape[real_axis] will always be num_outputs.

Comment (sxjscience) [inline src/operator/slice_channel-inl.h]: OK, I find that the current logic is that if (squeeze_axis=true && shape becomes 1), squeeze the dimension... However it's unnecessary to use squeeze_axis if the ndim is unchanged.

Comment (piiswrong) [inline src/operator/slice_channel-inl.h]: Sure. then let's add a check for that and update the document to force that behavior

**Pull request merged**